src/l2/lower_mac.cpp

```diff
@@ -34,126 +48,83 @@ auto LowerMac::process(const std::vector<uint8_t>& frame, BurstType burst_type)
-        // SynchronizationBurst is useless. However, the following SynchronizationBurst are decoded
-        // correctly since the scrambling code could, but proprobably will not change.
-        descramble(frame.data() + 252, bb_desc, 30, upper_mac_->scrambling_code());
+        descramble(frame.data() + 252, bb_desc, 30, bsc.scrambling_code);
         std::vector<uint8_t> bb_rm(14);
         reed_muller_3014_decode(bb_desc, bb_rm.data());
-        functions.push_back(std::bind(&UpperMac::process_AACH, upper_mac_, burst_type, bb_rm));
+        auto _aach = AccessAssignmentChannel(burst_type, bsc.time, bb_rm);
 
         // bkn2 block
         // ✅ done
         // any off SCH/HD, BNCH, STCH
         // see ETSI EN 300 392-2 V3.8.1 (2016-08) Figure 8.6: Error control
         // structure for π4DQPSK logical channels (part 2)
         uint8_t bkn2_desc[216];
-        // XXX: upper_mac_->scrambling_code() may yield the wrong result!!! if the processing of BSCH
-        // is not done before this call, and it cannot! This means the following data of the first
-        // SynchronizationBurst is useless. However, the following SynchronizationBurst are decoded
-        // correctly since the scrambling code could, but proprobably will not change.
-        descramble(frame.data() + 282, bkn2_desc, 216, upper_mac_->scrambling_code());
+        descramble(frame.data() + 282, bkn2_desc, 216, bsc.scrambling_code);
         uint8_t bkn2_dei[216];
         deinterleave(bkn2_desc, bkn2_dei, 216, 101);
         bkn2 = viter_bi_decode_1614(depuncture23(bkn2_dei, 216));
-        // if the crc does not work, then it might be a BLCH
         if (check_crc_16_ccitt(bkn2.data(), 140)) {
             bkn2 = std::vector(bkn2.begin(), bkn2.begin() + 124);
 
             // SCH/HD or BNCH mapped
-            functions.push_back(std::bind(&UpperMac::process_SCH_HD, upper_mac_, burst_type, bkn2));
+            functions.emplace_back(std::bind(&UpperMac::process_SCH_HD, upper_mac_, burst_type, bkn2));
         }
     } else if (burst_type == BurstType::NormalDownlinkBurst) {
-        upper_mac_->incrementTn();
-
         // bb contains AACH
         // ✅ done
         std::vector<uint8_t> bb{};
         vectorAppend(frame, bb, 230, 14);
         vectorAppend(frame, bb, 266, 16);
         uint8_t bb_des[30];
-        // XXX: upper_mac_->scrambling_code() may yield the wrong result!!! if the processing of the
-        // SynchronizationBurst is not done before this call, this may mean that the first packets are
-        // not decoded correctly.
-        descramble(bb.data(), bb_des, 30, upper_mac_->scrambling_code());
+        descramble(bb.data(), bb_des, 30, bsc.scrambling_code);
         std::vector<uint8_t> bb_rm(14);
         reed_muller_3014_decode(bb_des, bb_rm.data());
-        functions.push_back(std::bind(&UpperMac::process_AACH, upper_mac_, burst_type, bb_rm));
+        auto aach = AccessAssignmentChannel(burst_type, bsc.time, bb_rm);
```

Comment:
can we have a nicer name?

Reply:
well we can try, but sadly these are the ones which are used in the standard

Reply:
it just makes it incredibly hard to read that you always need to back to check what the different variables mean